src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestUpdate.java

```diff
@@ -0,0 +1,2161 @@
+
+        String tagsBefore = w.launchCommand("git", "tag");
+        Set<String> tagNamesBefore = w.git.getTagNames(null);
+        for (String tag : tagNamesBefore) {
+            assertTrue(tag + " not in " + tagsBefore, tagsBefore.contains(tag));
+        }
+
+        w.git.checkout().branch("tests/getSubmodules").ref("origin/tests/getSubmodules").timeout(checkoutTimeout).execute();
+        w.git.submoduleUpdate().recursive(true).execute();
+
+        String tagsAfter = w.launchCommand("git", "tag");
+        Set<String> tagNamesAfter = w.git.getTagNames(null);
+        for (String tag : tagNamesAfter) {
+            assertTrue(tag + " not in " + tagsAfter, tagsAfter.contains(tag));
+        }
+
+        assertEquals("tags before != after", tagsBefore, tagsAfter);
+
+        GitClient gitNtp = w.git.subGit("modules/ntp");
+        Set<String> tagNamesSubmodule = gitNtp.getTagNames(null);
+        for (String tag : tagNamesSubmodule) {
+            assertFalse("Submodule tag " + tag + " in parent " + tagsAfter, tagsAfter.matches("^" + tag + "$"));
+        }
+
+        try {
+            w.igit().fixSubmoduleUrls("origin", listener);
+            assertTrue("not CliGit", w.igit() instanceof CliGitAPIImpl);
+        } catch (UnsupportedOperationException uoe) {
+            assertTrue("Unsupported operation not on JGit", w.igit() instanceof JGitAPIImpl);
+        }
+    }
+
+
+    /* Shows the JGit submodule update is broken now that tests/getSubmodule includes a renamed submodule */
+    @Test
+    public void testGetSubmodules() throws Exception {
+        w.init();
+        w.git.clone_().url(localMirror()).repositoryName("sub_origin").execute();
+        w.git.checkout().ref("sub_origin/tests/getSubmodules").branch("tests/getSubmodules").execute();
+        List<IndexEntry> r = w.git.getSubmodules("HEAD");
+        assertEquals(
+                "[IndexEntry[mode=160000,type=commit,file=modules/firewall,object=978c8b223b33e203a5c766ecf79704a5ea9b35c8], " +
+                        "IndexEntry[mode=160000,type=commit,file=modules/ntp,object=b62fabbc2bb37908c44ded233e0f4bf479e45609], " +
+                        "IndexEntry[mode=160000,type=commit,file=modules/sshkeys,object=689c45ed57f0829735f9a2b16760c14236fe21d9]]",
+                r.toString()
+        );
+        w.git.submoduleInit();
+        w.git.submoduleUpdate().execute();
+
+        assertTrue("modules/firewall does not exist", w.exists("modules/firewall"));
+        assertTrue("modules/ntp does not exist", w.exists("modules/ntp"));
+        // JGit submodule implementation doesn't handle renamed submodules
+        if (w.igit() instanceof CliGitAPIImpl) {
+            assertTrue("modules/sshkeys does not exist", w.exists("modules/sshkeys"));
+        }
+        assertFixSubmoduleUrlsThrows();
+    }
+
+
+    /* Shows the submodule update is broken now that tests/getSubmodule includes a renamed submodule */
+    @NotImplementedInJGit
```

Comment:
This annotation is not honored in the implementation you've created for JUnit 4.  It was used in JUnit 3 to cause the test to be skipped when the implementation was JGit.  In JUnit 4 you can either move the test into a class that is dedicated to testing the CliGit implementation or you can add a conditional at the start of the test that exits immediately if the git implementation is not command line git.

I prefer that it be moved to a separate file so that the test count is not incorrectly suggesting that this test is run for JGit.
```suggestion
```